int32 lfw
int32 lmw
int32 lrw
int32 rfw
int32 rmw
int32 rrw
